File: RQ1/CodeForces-Automatic/158B.c
#include <stdio.h>
#include <assert.h>

int main()
{
    int n;
    scanf("%d", &n);
    assert(n >= 1 && n <= 100000); // PRE: 1 ≤ n ≤ 10^5

    int ct1 = 0, ct2 = 0, ct3 = 0, ct4 = 0;
    for (int i = 0; i < n; ++i)
    {
        int x;
        scanf("%d", &x);
        assert(x >= 1 && x <= 4); // PRE: 1 ≤ si ≤ 4
        if (x == 1)
            ct1++;
        else if (x == 2)
            ct2++;
        else if (x == 3)
            ct3++;
        else
            ct4++;
    }
    
    int taxi = ct4;
    if (ct2 > 1)
    {
        taxi += ct2 / 2;
        ct2 = ct2 % 2;
    }
    while (ct3 > 0 && ct1 > 0)
    {
        taxi++;
        ct3--;
        ct1--;
    }
    if (ct3 > 0)
        taxi += ct3;
    if (ct1 > 0 && ct2 > 0)
    {
        taxi++;
        ct2--;
        ct1 -= 2;
    }
    while (ct2 > 0)
    {
        taxi++;
        ct2--;
    }
    while (ct1 > 0)
    {
        taxi++;
        ct1 -= 4;
    }
    
    printf("%d\n", taxi);
    assert(taxi >= (ct4 + (ct3 + 1) / 2 + (ct1 + 3) / 4 + (ct2 + 1) / 2)); // POST: minimum taxis calculated correctly
    return 0;
}

File: RQ1/CodeForces-Automatic/158B.py
def main():
    # PRE:
    # Input Format:
    # The first line contains integer n (1 ≤ n ≤ 105) — the number of groups of schoolchildren. The second line contains a sequence of integers s1, s2, ..., sn (1 ≤ si ≤ 4). The integers are separated by a space, si is the number of children in the i-th group.
    # POST:
    # Program Description:
    # After the lessons n groups of schoolchildren went outside and decided to visit Polycarpus to celebrate his birthday. We know that the i-th group consists of si friends (1 ≤ si ≤ 4), and they want to go to Polycarpus together. They decided to get there by taxi. Each car can carry at most four passengers. What minimum number of cars will the children need if all members of each group should ride in the same taxi (but one taxi can take more than one group)?
    # Output Format:
    # Print the single number — the minimum number of taxis necessary to drive all children to Polycarpus.
    n = int(input())
    groups = input()
    
    count = {}
    count[1] = groups.count('1')
    count[2] = groups.count('2')
    count[3] = groups.count('3')
    count[4] = groups.count('4')
    
    total = count[4] + count[3] + count[2] // 2
    count[1] -= count[3]
    
    if (count[2] % 2 == 1):
          total += 1;
          count[1] -= 2;
    
    if (count[1] > 0):
          total += (count[1] + 3) // 4;
    
    print(total)
